File: Graph Algorithms/BFS/LeetCode - 542. 01 Matrix/ZeroOneMatrix.java
import java.util.LinkedList;
import java.util.Queue;

//     Java
// Time complexity - O(rows * cols)
// Space complexity - O(rows * cols)

public class ZeroOneMatrix {

    public int[][] updateMatrix(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return matrix;
        }

        int rows = matrix.length;
        int cols = matrix[0].length;

        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == 0) {
                    queue.offer(new int[]{i, j});
                    visited[i][j] = true;
                }
            }
        }

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0];
            int col = cell[1];

            for (int[] dir : dirs) {
                int newRow = row + dir[0];
                int newCol = col + dir[1];

                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols && !visited[newRow][newCol]) {
                    matrix[newRow][newCol] = matrix[row][col] + 1;
                    visited[newRow][newCol] = true;
                    queue.offer(new int[]{newRow, newCol});
                }
            }
        }

        return matrix;
    }
}

File: python/zero_one_matrix.py
from collections import deque

#     Python
# Time complexity - O(rows * cols)
# Space complexity - O(rows * cols)
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        if not matrix or not matrix[0]:
            return matrix

        rows, cols = len(matrix), len(matrix[0])
        queue = deque()
        visited = [[False] * cols for _ in range(rows)]

        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == 0:
                    queue.append((i, j))
                    visited[i][j] = True

        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        while queue:
            row, col = queue.popleft()
            for dx, dy in dirs:
                new_row, new_col = row + dx, col + dy
                if 0 <= new_row < rows and 0 <= new_col < cols and not visited[new_row][new_col]:
                    matrix[new_row][new_col] = matrix[row][col] + 1
                    visited[new_row][new_col] = True
                    queue.append((new_row, new_col))

        return matrix
